clients/adapter-amqp/src/main/java/org/eclipse/hono/adapter/client/registry/amqp/ProtonBasedDeviceRegistrationClient.java

```diff
@@ -0,0 +1,129 @@
+/**
+ * Copyright (c) 2020 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+
+
+package org.eclipse.hono.adapter.client.registry.amqp;
+
+import java.net.HttpURLConnection;
+import java.util.Objects;
+
+import org.eclipse.hono.adapter.client.registry.DeviceRegistrationClient;
+import org.eclipse.hono.cache.CacheProvider;
+import org.eclipse.hono.client.HonoConnection;
+import org.eclipse.hono.client.RegistrationClient;
+import org.eclipse.hono.client.SendMessageSampler;
+import org.eclipse.hono.client.ServerErrorException;
+import org.eclipse.hono.client.impl.CachingClientFactory;
+import org.eclipse.hono.client.impl.RegistrationClientImpl;
+import org.eclipse.hono.config.ProtocolAdapterProperties;
+import org.eclipse.hono.util.Constants;
+import org.eclipse.hono.util.RegistrationAssertion;
+import org.eclipse.hono.util.RegistrationConstants;
+import org.eclipse.hono.util.RegistrationResult;
+
+import io.opentracing.SpanContext;
+import io.vertx.core.Future;
+import io.vertx.core.eventbus.Message;
+import io.vertx.core.json.DecodeException;
+
+
+/**
+ * A vertx-proton based client of Hono's Device Registration service.
+ *
+ */
+public class ProtonBasedDeviceRegistrationClient extends AbstractRequestResponseClient<RegistrationResult>
+        implements DeviceRegistrationClient {
+
+    private final CachingClientFactory<org.eclipse.hono.client.RegistrationClient> clientFactory;
+
+    /**
+     * Creates a new client for a connection.
+     *
+     * @param connection The connection to the Device Registration service.
+     * @param samplerFactory The factory for creating samplers for tracing AMQP messages being sent.
+     * @param adapterConfig The protocol adapter's configuration properties.
+     * @param cacheProvider The cache provider to use for creating a cache for service responses.
+     * @throws NullPointerException if any of the parameters are {@code null}.
```

Comment:
`cacheProvider` is not being null checked. How about `@throws NullPointerException if any of the parameters other than cacheProvider are {@code null}.`?